File: 2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.py
MOD = 1000000007
class Solution:
    def numberOfPaths(self, grid: List[List[int]], k: int) -> int:
        n, m = len(grid), len(grid[0])
        dp = [[[0] * k for _ in range(m+1)] for _ in range(n+1)]
        dp[1][1][grid[0][0] % k] = 1
        for i in range(n):
            for j in range(m):
                if i == 0 and j == 0: continue
                v = grid[i][j] % k
                for q in range(k):
                    f, s = dp[i+1][j][(k+q-v) % k], dp[i][j+1][(k+q-v) % k]
                    # print(i,j,q,v,f,s,k+q-v)
                    dp[i+1][j+1][q] = ((f) + (s)) % MOD
        # for r in dp:
        #     print(r)
        return dp[-1][-1][0]

File: 2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.cpp
class Solution {
private:
    int MOD = 1000000007;
public:
    int numberOfPaths(vector<vector<int>>& grid, int k) {
        int n = grid.size();
        int m = grid[0].size();
        if (n == 1 || m == 1) {
            int ts = 0;
            for (auto& row: grid) {
                for (int x: row) {
                    ts += x;
                }
            }
            return (ts % k == 0) ? 1 : 0;
        }
        vector<vector<vector<int>>> dp(n, vector<vector<int>>(m, vector<int>(k, 0)));;
        dp[0][0][grid[0][0] % k] = 1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (i == 0 && j == 0) continue;
                int v = grid[i][j] % k;
                for (int q = 0; q < k; q++) {
                    int r = (q + v) % k;
                    int res = 0;
                    if (i > 0) {
                        res += dp[i-1][j][q];
                    }
                    if (j > 0) {
                        res += dp[i][j-1][q];
                    }
                    dp[i][j][r] = (dp[i][j][r] + res) % MOD;
                }
            }
        }
        return dp[n-1][m-1][0];
    }
};